broadcast/src/linked/signer/actor.rs

```diff
@@ -0,0 +1,1073 @@
+    }
+
+    ////////////////////////////////////////
+    // Journal
+    ////////////////////////////////////////
+
+    /// Returns the section of the journal for the given height.
+    fn get_journal_section(&self, height: u64) -> u64 {
+        height / self.journal_heights_per_section
+    }
+
+    /// Ensures the journal exists and is initialized for the given sequencer.
+    /// If the journal does not exist, it is created and replayed.
+    /// Else, no action is taken.
+    async fn journal_prepare(&mut self, sequencer: &C::PublicKey) {
+        // Return early if the journal already exists
+        if self.journals.contains_key(sequencer) {
+            return;
+        }
+
+        // Initialize journal
+        let cfg = journal::variable::Config {
+            registry: Arc::new(Mutex::new(Registry::default())),
+            partition: (self.journal_naming_fn)(sequencer),
+        };
+        let mut journal = Journal::init(self.runtime.clone(), cfg)
+            .await
+            .expect("unable to init journal");
+
+        // Replay journal
+        {
+            debug!(?sequencer, "journal replay begin");
+
+            // Prepare the stream
+            let stream = journal
+                .replay(self.journal_replay_concurrency, None)
+                .await
+                .expect("unable to replay journal");
+            pin_mut!(stream);
+
+            // Read from the stream, which may be in arbitrary order.
+            // Remember the highest link height
+            let mut tip: Option<safe::Link<C, D>> = None;
+            let mut num_items = 0;
+            while let Some(msg) = stream.next().await {
+                num_items += 1;
+                let (_, _, _, msg) = msg.expect("unable to decode journal message");
+                let link =
+                    safe::Link::<C, D>::decode(&msg).expect("journal message is unexpected format");
+                let height = link.chunk.height;
+                match tip {
+                    None => {
+                        tip = Some(link);
+                    }
+                    Some(ref t) => {
+                        if height > t.chunk.height {
+                            tip = Some(link);
+                        }
+                    }
+                }
+            }
+
+            // Set the tip only once. The items from the journal may be in arbitrary order,
+            // and the tip manager will panic if inserting tips out-of-order.
+            if let Some(link) = tip.take() {
+                let is_new = self.tip_manager.put(&link);
+                assert!(is_new);
+            }
+
+            debug!(?sequencer, ?num_items, "journal replay end");
+        }
+
+        // Store journal
+        self.journals.insert(sequencer.clone(), journal);
+    }
+
+    /// Write a link to the appropriate journal.
+    ///
+    /// The journal must already be open and replayed.
+    async fn journal_append(&mut self, link: &safe::Link<C, D>) {
```

Comment:
I keep forgetting that `Link` includes the new chunk (and get scared we can double-sign) when reading this. I wonder if it makes sense to include a `(..)` somewhere being explicit that this includes the chunk at a given height.